ral-facilities/datagateway-api
# Implement Include Filter for Search API

This PR will close #261 

## Description
This PR implements the include filter for the search API. The logic to this is fairly simple and similar to the where filter actually - convert the PaNOSC field names to ICAT field names and then let Python ICAT add the relevant stuff to the JPQL query (via the code in `PythonICATIncludeFilter`).

I've moved `get_icat_mapping()` into `PaNOSCMappings` as it seems to be more logically placed there. In doing so, I removed the logic that deals with the special use case where the ICAT mapping is a list of field names and put it in `apply_filter()`. I've done this to keep `get_icat_mapping()` a bit more generic, though we can move it back at a later stage if needed (if we find that the logic needs to be executed in multiple places). 

I have also refactored the query params code. The include filter can support a list of related fields, but the query params only created include filters for a single related entity so multiple `SearchAPIIncludeFilter`s would be created when only one is necessary. I've made the query params able to create include filters with multiple related fields (as proved by the diff of `test_search_api_query_filter_factory.py` in eab7f3d9d10eb36a998653a492c99476fc7f46d9).

## Testing Instructions
I guess similar to #298, the method of manual testing would be to send some requests to `/search-api/datasets` with include filters and see what you can break. If the filter is successfully created, the response body will show the version of ICAT server which is just a placeholder until the endpoint is actually implemented. As above, not all the mappings will work because some need ICAT 5.0 (currently unreleased) but I've added them in so we don't forget about them.

There is a feature version bump. I'm unsure of the exact numeric version bump there'll be because we have a few PRs awaiting to be merged but `poetry run semantic-release print-version` shows that the version bump does work (currently bumped to 3.2.0 as the latest version is 3.1.1). 


## Agile Board Tracking
Connect to #261
